proposed/MPID Messaging.md

```diff
@@ -0,0 +1,295 @@
+
+## Supported Use-Cases
+
+The system will support secure, private, bi-directional communications between pairs of Clients.
+
+## Expected Outcome
+
+The provision of a secure messaging system which will eradicate unwanted and intrusive communications.
+
+# Detailed design
+
+## Overview
+
+A fundamental principle is that the cost of messaging is with the sender, primarily to deter unsolicited messages.  To achieve this, the sender will maintain messages in a network [outbox][3] until they are retrieved by the recipient.  The recipient will have a network [inbox][4] comprising a list of metadata relating to messages which are trying to be delivered to it.
+
+If the message is unwanted, the recipient simply does not retrieve the message.  The sender will thus quickly fill their own outbox with undelivered mail and be forced to clean this up themselves before being able to send further messages.
+
+This paradigm shift will mean that the obligation to unsubscribe from mailing lists, etc. is now with the owner of these lists.  If people are not picking up messages, it is because they do not want them.  So the sender has to do a better job.  It is assumed this shift of responsibilities will lead to a better-managed bandwidth solution and considerably less stress on the network and the users of the network.
+
+## Implementation Details
+
+The two relevant structs (other than the MPID itself which is a [standard Client key][0]) are the [`MpidHeader`][1] and the [`MpidMessage`][2].
+
+Broadly speaking, the `MpidHeader` contains metadata and the `MpidMessage` contains a signed header and message.  We also want to define some upper limits which will be described later.
+
+### Consts
+
+```rust
+pub const MPID_MESSAGE_TAG: u64 = 51000;
+pub const MPID_HEADER_TAG: u64 = 51001;
+pub const MAX_SUBJECT_SIZE: usize = 128;  // bytes
+pub const MAX_INBOX_SIZE: usize = 1 << 27;  // bytes, i.e. 128 MiB
+pub const MAX_OUTBOX_SIZE: usize = 1 << 27;  // bytes, i.e. 128 MiB
+```
+
+### `MpidHeader`
+
+```rust
+pub struct MpidHeader {
+    pub sender: ::routing::NameType,
+    pub recipient: ::routing::NameType,
+    pub index: u32,
+    pub parent_index: Option<u32>,
+    pub subject: Vec<u8>,
+}
+```
+
+The `sender` and `recipient` fields are hopefully self-explanatory.  However, the only purpose for them existing in the `MpidHeader` is to increase the randomness of the derived message name so conflicts have a much lower chance of happening.
+
+The `index` and `parent_index` are intended to allow threading or sequencing of messages, while `subject` allows passing arbitrary data (e.g. a subject line).
+
+The `subject` field must not exceed `MAX_SUBJECT_SIZE` bytes.
+
+### `MpidMessage`
+
+```rust
+pub struct MpidMessage {
```

Comment:
Ok .. It makes sense in combination with other answer:
> The clients will then have to retrieve the corresponding MPID public keys from the network and presumably hold these in some form of address book for the user

So will this be the verification flow ? :
1. Receiver Client (`B`) stores in address-book mapping of `sender name -> sender MPID Public Sign Key`
1. Since from the discussions, it seems `pull` model is heavily favored, `B` does a `GET` for the messages in Inbox and gets `(sender name <-> signed, also probably encrypted, message_header)`.
1. `B` obtains `sender MPID Public Sign Key` from its address book and verifies the received message_header, similarly proceeds to `GET` message_body.
1. If not present in `B`'s address-book then there is no way B can verify the signature and then it's `B`'s call whether to act on it or discard it.

Is that right ? ^^^